### ISSUE TITLE
Texture broken with mipmaps enabled

### ISSUE THREAD

**Issue by KoBeWi** [godotengine/godot#50647]
### Godot version

3.4 beta1

### System information

W10

### Issue description

I have a texture that can't be imported with mipmaps enabled. It shows this error:
```
 Expected data size of 1528 bytes in Image::create(), got instead 1527 bytes.
 drivers/gles3/rasterizer_storage_gles3.cpp:757 - Condition "!read.ptr()" is true.
```
And image format shows Lum8. The texture becomes black rectangle.
This doesn't happen with other textures, even smaller ones.

Apparently it was fine some time ago and only broke recently. Also works in 4.0.

### Steps to reproduce

1. Copy this to your project
![PistolBullet](https://user-images.githubusercontent.com/2223172/126333679-2cfccac7-62a3-413d-9966-489d6b3f8e1f.png)
2. Enable Mipmaps in import options

**Comment by akien-mga** [comment]
Confirmed on Jetpaca with this texture: https://github.com/KOBUGE-Games/jetpaca/blob/master/art/intro_sky_bg.png

It's a regression specific to `3.x` (present in 3.4-beta1 and latest `3.x` commit).

**Comment by akien-mga** [comment]
Had a couple older builds lying around, the bug is not reproduced in a69cc9f13da50ddf18452cfac2a5755624e1e971 (May 22), but it is reproducible in 83ad0dddad5da10f4917de9902c13aace2a9ae68 (July 8).

Time to bisect...

**Comment by akien-mga** [comment]
Bisected to #47854, CC @mortarroad.

**Comment by mortarroad** [comment]
Hmm, doesn't seem to be caused by my commit, just revealed by it.
The WebP unpacker returns its images as either RGB or RGBA, depending on whether the decoded image actually has transparent pixels.

However, the Texture resource loaded assumes all of its images to have the same format. It just uses the format of the first mipmap image, and copies them consecutively in memory.
https://github.com/godotengine/godot/blob/ab30437529ef0af6a96e4f2a8a92a050a09dbca0/scene/resources/texture.cpp#L582-L599

Different formats can happen, if the source image has very few transparent pixels, such that a smaller mip map ends up with no transparency.
This issue also affects the lossy WebP encoder.

A quick fix is to let WebP always decode to RGBA.
Another is to properly convert all mipmap images to the same format, before combining them.

**Comment by mortarroad** [comment]
Adding this
```
			if (i != 0) {
				img->convert(mipmap_images[0]->get_format()); // ensure the same format for all mipmaps
			}
```

right here https://github.com/godotengine/godot/blob/ab30437529ef0af6a96e4f2a8a92a050a09dbca0/scene/resources/texture.cpp#L567 (i.e. right _before_ the line with the `total_size`) fixes it.


Also, while I'm here, let me reiterate how silly it is to store mipmaps on disk other than for VRAM compression formats.

**Comment by mortarroad** [comment]
Looking into 4.x, that seems to have equivalent code.
See:
https://github.com/godotengine/godot/blob/7f0a83c1291bad864fc094bec8b763986172e73e/scene/resources/texture.cpp#L369-L376

My fix is shorter though, hehe.

**Comment by reduz** [comment]
Converting on load is inefficient, IMO the webp decoder should be told the desired format instead with a hint.

**Comment by mortarroad** [comment]
I disagree. I estimate that a conversion happens in less than 1% of the cases and then only for the lowest mipmap levels, which are small anyway (a few pixels at worst).
However, it could make sense to store the used channels anyway, and use then when loading to save memory. That would be a new feature, however.
(Or even better, we don't store mipmaps on disk, and compute them on the fly.)

Either way, the proper fix here is to add the conversion, as in 4.x.

Looking through the code, this issue affects at least one more encoder, possibly others too:
The png encoder also encodes without alpha, if the source image has a more exotic channel layout:
https://github.com/godotengine/godot/blob/0543b71e5a7c848e7eec233c428fb2c1fb5caca8/drivers/png/png_driver_common.cpp#L155-L162

**Comment by reduz** [comment]
Yeah upon closer inspection you are right, this should happen so little often its not worth having a more complex workaround.

**Comment by akien-mga** [comment]
Fixed by #51555.
